backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/ChangeLogUtils.java

```diff
@@ -12,33 +17,96 @@ private ChangeLogUtils() {
-                   END;
-                   $%3$s_audit$ LANGUAGE plpgsql;
-                """
+        CREATE OR REPLACE FUNCTION "%1$s"."process_%3$s_audit"()
+        RETURNS TRIGGER AS $%3$s_audit$
+        DECLARE
+            column_names varchar[] := ARRAY[%4$s];
+            old_row JSONB;
+            new_row JSONB;
+            col_name TEXT;
+            old_value TEXT;
+            new_value TEXT;
+        BEGIN
+            -- Initialize empty JSONB objects
+            old_row := '{}'::JSONB;
+            new_row := '{}'::JSONB;
+
+            -- Loop through each column in the OLD record
+            FOREACH col_name IN ARRAY column_names
+            LOOP
+                -- Skip columns that end with '_contents' or '_TEXT_SEARCH_COLUMN'
+                IF col_name LIKE '%%_contents' OR col_name LIKE '%%_TEXT_SEARCH_COLUMN' OR col_name LIKE 'mg_%%' THEN
+                    CONTINUE;
+                END IF;
+                IF TG_OP != 'INSERT' THEN
+                    EXECUTE 'SELECT ($1).' || quote_ident(col_name) INTO old_value USING OLD;
+                    IF old_value IS NOT NULL THEN
+                      old_row := jsonb_set(old_row, ARRAY[col_name], to_jsonb(old_value::TEXT)::JSONB);
+                    END IF;
+                END IF;
+                IF TG_OP != 'DELETE' THEN
+                    EXECUTE 'SELECT ($1).' || quote_ident(col_name) INTO new_value USING NEW;
+                    IF new_value IS NOT NULL THEN
+                      new_row := jsonb_set(new_row, ARRAY[col_name], to_jsonb(new_value::TEXT)::JSONB);
+                    END IF;
+                END IF;
+            END LOOP;
+
+            -- Log the change based on the operation
+            IF TG_OP = 'DELETE' THEN
+                INSERT INTO "%1$s".mg_changelog
+                SELECT 'D', now(), user, TG_TABLE_NAME, old_row, new_row;
+            ELSIF TG_OP = 'UPDATE' THEN
+                INSERT INTO "%1$s".mg_changelog
+                SELECT 'U', now(), user, TG_TABLE_NAME, old_row, new_row;
+            ELSIF TG_OP = 'INSERT' THEN
+                INSERT INTO "%1$s".mg_changelog
+                SELECT 'I', now(), user, TG_TABLE_NAME, old_row, new_row;
+            END IF;
+
+            RETURN NULL; -- result is ignored since this is an AFTER trigger
+        END;
+        $%3$s_audit$ LANGUAGE plpgsql;
+        """
         .formatted(
-            schemaName,
-            ChangeLogUtils.buildFunctionName(schemaName),
-            ChangeLogUtils.buildFunctionName(tableName));
+            tableMetadata.getSchemaName(),
+            ChangeLogUtils.buildFunctionName(tableMetadata.getSchemaName()),
+            ChangeLogUtils.buildFunctionName(tableMetadata.getTableName()),
+            columnNameArray);
   }
 
   public static String buildAuditTrigger(String schemaName, String tableName) {
     return """
-            CREATE TRIGGER %3$s_audit
+            CREATE OR REPLACE TRIGGER %3$s_audit
```

Comment:
???

Reply:
ah, with this change updates can happen because this currently uses same code path as a trigger function update => when data model changes the trigger function should also be updated to accomodate column changes). I cannot be bothered to split it, however we could refactor it to also include the updating of the search indexing function in some future.

Reply:
yup. It is a drag but I think running out of storage because a record containing large images is often updated I believe is a much bigger risk